source/extensions/filters/network/ratelimit/ratelimit.h

```diff
@@ -83,13 +85,44 @@ class Filter : public Network::ReadFilter,
   void initializeReadFilterCallbacks(Network::ReadFilterCallbacks& callbacks) override {
     filter_callbacks_ = &callbacks;
     filter_callbacks_->connection().addConnectionCallbacks(*this);
+    applySubstitutionFormatter(config_->descriptors(),
+                               filter_callbacks_->connection().streamInfo());
+  }
+
+  void applySubstitutionFormatter(std::vector<RateLimit::Descriptor> original_descriptors,
+                                  StreamInfo::StreamInfo& stream_info) {
+
+    std::vector<RateLimit::Descriptor> dynamicDescriptors = std::vector<RateLimit::Descriptor>();
+    for (const RateLimit::Descriptor& descriptor : original_descriptors) {
+      RateLimit::Descriptor new_descriptor;
+      for (const RateLimit::DescriptorEntry& descriptorEntry : descriptor.entries_) {
+        std::string value = descriptorEntry.value_;
+        value = formatValue(value, stream_info);
+        new_descriptor.entries_.push_back({descriptorEntry.key_, value});
+      }
+      dynamicDescriptors.push_back(new_descriptor);
+    }
+    filter_descriptors_ = dynamicDescriptors;
+  }
+
+  std::string formatValue(std::string body, StreamInfo::StreamInfo& stream_info) {
+    Http::RequestHeaderMapPtr request_headers = Http::RequestHeaderMapImpl::create();
+    Http::ResponseHeaderMapPtr response_headers = Http::ResponseHeaderMapImpl::create();
+    Http::ResponseTrailerMapPtr response_trailers = Http::ResponseTrailerMapImpl::create();
+
+    Formatter::FormatterImpl formatter(body);
+
+    std::string value = formatter.format(*request_headers.get(), *response_headers.get(),
```

Comment:
can this also implement on http ratelimit?

Reply:
Did you mean can we implement this in the HTTP rate limiter?

Reply:
Yes, optimally we would do this change in both places, but we could do this in a follow up.

Reply:
I'd prefer to do that in a followup. I can do that after this one.